File: kentender_core/kentender_core/public/js/reference_data/data/referenceDataApi.js
// Thin frappe.call wrappers over kentender_core.api.reference_data_api. No business
// logic here — every guard (permission, state, SoD, idempotency) lives server-side;
// this module only shapes requests/responses for the composables.

const METHOD_PREFIX = "kentender_core.api.reference_data_api.";

// Parses Frappe's raw error payload (from frappe.call's error callback) into
// {code, message}, where `code` is the server's frappe.throw(..., title=...)
// value when set. `silent:true` below stops frappe.call's default behaviour of
// popping its own generic error dialog — screens handle {code, message} themselves.
function parseServerError(r) {
	let messages = [];
	try {
		if (r && r._server_messages) {
			messages = JSON.parse(r._server_messages).map((m) => (typeof m === "string" ? JSON.parse(m) : m));
		}
	} catch (e) {
		/* fall through to the generic message below */
	}
	const first = messages[0] || {};
	return {
		code: first.title || (r && r.exc_type) || null,
		message: first.message || __("Something went wrong. Please try again."),
		excType: (r && r.exc_type) || null,
	};
}

function call(method, args) {
	return new Promise((resolve, reject) => {
		frappe.call({
			method: METHOD_PREFIX + method,
			args,
			silent: true,
			callback: (r) => resolve(r.message),
			error: (r) => reject(parseServerError(r)),
		});
	});
}

function newIdempotencyKey() {
	return `rd-${Date.now()}-${Math.random().toString(36).slice(2, 10)}`;
}

export const referenceDataApi = {
	listPeTypes: () => call("list_pe_types", {}),

	// --- Procuring Entity ---
	listProcuringEntities: (params = {}) => call("list_procuring_entities", params),
	getProcuringEntity: (peId) => call("get_procuring_entity", { pe_id: peId }),
	createPe: (payload) => call("create_or_revise_pe", { payload }),
	updatePeDraft: (peId, payload) => call("update_pe_draft", { pe_id: peId, payload }),
	proposePeAmendment: (peId, changeReason) =>
		call("create_or_revise_pe", { pe_id: peId, change_reason: changeReason }),
	decidePeChange: (peId, action, extra = {}) =>
		call("decide_pe_change", { pe_id: peId, action, idempotency_key: newIdempotencyKey(), ...extra }),

	// --- Financial Year ---
	listFinancialYears: (params = {}) => call("list_financial_years", params),
	getFinancialYear: (financialYearId) => call("get_financial_year", { financial_year_id: financialYearId }),
	createFinancialYear: (startYear) => call("create_financial_year", { start_year: startYear }),
	makeFinancialYearAvailable: (financialYearId) =>
		call("make_financial_year_available", { financial_year_id: financialYearId, idempotency_key: newIdempotencyKey() }),
	retireFinancialYear: (financialYearId) =>
		call("retire_financial_year", { financial_year_id: financialYearId, idempotency_key: newIdempotencyKey() }),

	// --- PE Fiscal Year Context ---
	listPeFyContexts: (params = {}) => call("list_pe_fy_contexts", params),
	getPeFyContext: (contextId) => call("get_pe_fy_context", { context_id: contextId }),
	enablePeFyContext: (procuringEntity, financialYear, activeFrom, activeTo) =>
		call("enable_pe_fy_context", {
			procuring_entity: procuringEntity,
			financial_year: financialYear,
			active_from: activeFrom,
			active_to: activeTo,
			idempotency_key: newIdempotencyKey(),
		}),
	decidePeFyContext: (contextId, action, expectedVersion, extra = {}) =>
		call("decide_pe_fy_context", {
			context_id: contextId,
			action,
			idempotency_key: newIdempotencyKey(),
			expected_version: expectedVersion,
			...extra,
		}),
};

File: kentender_core/kentender_core/api/reference_data_api.py
# Copyright (c) 2026, KenTender and contributors
# For license information, please see license.txt

"""CFG-CHG-002 §10 — whitelisted service contracts. Thin wrappers only: every
guard (permission, state, SoD, generated-field immutability) lives in the
services layer (reference_data_transitions/reference_data_resolver); this
module validates presence of required input, dispatches by action, and wires
idempotency where the contract calls for it — it enforces nothing itself.

A few endpoints beyond the 12 named in §10 are exposed as pragmatic, necessary
supplements: the FY/Context state machines have actor-gated steps (e.g. FY's
own Submit, Context's Recommend) that the terse contract table doesn't name
separately from Create/Decide but that a real UI must still be able to call.
Each is marked below.
"""

from __future__ import annotations

import frappe

from kentender_core.services import reference_data_queries as queries
from kentender_core.services import reference_data_resolver as resolver
from kentender_core.services import reference_data_transitions as txn
from kentender_core.services.reference_data_idempotency import run_idempotent


def _obj(payload) -> dict:
	if isinstance(payload, str):
		return frappe.parse_json(payload) or {}
	return payload or {}


# --- PE Type -------------------------------------------------------------------------


@frappe.whitelist()
def list_pe_types():
	return queries.list_pe_types()


# --- Procuring Entity --------------------------------------------------------------


@frappe.whitelist()
def list_procuring_entities(status: str | None = None, pe_type: str | None = None, search: str | None = None):
	return queries.list_procuring_entities(status=status, pe_type=pe_type, search=search)


@frappe.whitelist()
def get_procuring_entity(pe_id: str):
	return queries.get_procuring_entity(pe_id)


@frappe.whitelist()
def create_or_revise_pe(payload=None, pe_id: str | None = None, change_reason: str | None = None):
	"""Create when pe_id is omitted; revise (propose amendment) when given."""
	if pe_id:
		return txn.propose_amendment(pe_id, change_reason or "", user=frappe.session.user)
	return txn.create_pe_draft(_obj(payload), user=frappe.session.user)


@frappe.whitelist()
def update_pe_draft(pe_id: str, payload=None):
	"""Edit a still-Draft PE's fields before Activate — §6.1's 'Create draft' and
	'Activate' are separate steps, so the draft must stay editable in between."""
	return txn.update_pe_draft(pe_id, _obj(payload), user=frappe.session.user)


@frappe.whitelist()
def decide_pe_change(
	pe_id: str,
	action: str,
	reason: str | None = None,
	effective_date=None,
	idempotency_key: str | None = None,
):
	dispatch = {
		"activate": lambda: txn.activate_pe(pe_id, user=frappe.session.user),
		"suspend": lambda: txn.suspend_pe(pe_id, reason or "", user=frappe.session.user),
		"reinstate": lambda: txn.reinstate_pe(pe_id, user=frappe.session.user),
		"retire": lambda: txn.retire_pe(
			pe_id, reason or "", effective_date or frappe.utils.today(), user=frappe.session.user
		),
	}
	if action not in dispatch:
		frappe.throw(f"Unknown Procuring Entity action: {action}", frappe.ValidationError)
	return run_idempotent(idempotency_key, "Procuring Entity", pe_id, action, dispatch[action])


# --- Financial Year -----------------------------------------------------------------


@frappe.whitelist()
def list_financial_years(record_status: str | None = None):
	return queries.list_financial_years(record_status=record_status)


@frappe.whitelist()
def get_financial_year(financial_year_id: str):
	"""Supplement — needed for §12.11's FY detail screen; not separately named in §10."""
	return queries.get_financial_year(financial_year_id)


@frappe.whitelist()
def create_financial_year(start_year):
	return txn.create_fy_draft(int(start_year), user=frappe.session.user)


@frappe.whitelist()
def make_financial_year_available(financial_year_id: str, idempotency_key: str | None = None):
	return run_idempotent(
		idempotency_key,
		"Financial Year",
		financial_year_id,
		"make_available",
		lambda: txn.make_fy_available(financial_year_id, user=frappe.session.user),
	)


@frappe.whitelist()
def retire_financial_year(financial_year_id: str, idempotency_key: str | None = None):
	"""Supplement — §10 doesn't name FY retirement separately; §6.2 has it."""
	return run_idempotent(
		idempotency_key,
		"Financial Year",
		financial_year_id,
		"retire",
		lambda: txn.retire_fy(financial_year_id, user=frappe.session.user),
	)


# --- PE Fiscal Year Context ----------------------------------------------------------


@frappe.whitelist()
def list_pe_fy_contexts(
	procuring_entity: str | None = None,
	financial_year: str | None = None,
	status: str | None = None,
	search: str | None = None,
):
	return queries.list_pe_fy_contexts(
		procuring_entity=procuring_entity, financial_year=financial_year, status=status, search=search
	)


@frappe.whitelist()
def get_pe_fy_context(context_id: str):
	return queries.get_pe_fy_context(context_id)


@frappe.whitelist()
def enable_pe_fy_context(
	procuring_entity: str,
	financial_year: str,
	active_from=None,
	active_to=None,
	idempotency_key: str | None = None,
):
	return run_idempotent(
		idempotency_key,
		"PE Fiscal Year Context",
		f"{procuring_entity}:{financial_year}",
		"enable",
		lambda: txn.enable_context(procuring_entity, financial_year, active_from, active_to, user=frappe.session.user),
	)


@frappe.whitelist()
def decide_pe_fy_context(
	context_id: str,
	action: str,
	reason: str | None = None,
	acknowledged=False,
	active_from=None,
	active_to=None,
	idempotency_key: str | None = None,
	expected_version: str | None = None,
):
	dispatch = {
		"suspend": lambda: txn.suspend_context(
			context_id, reason or "", user=frappe.session.user, expected_version=expected_version
		),
		"reinstate": lambda: txn.reinstate_context(
			context_id, user=frappe.session.user, expected_version=expected_version
		),
		"close": lambda: txn.close_context(
			context_id,
			reason or "",
			acknowledged=frappe.utils.cint(acknowledged) == 1,
			user=frappe.session.user,
			expected_version=expected_version,
		),
		"reopen": lambda: txn.reopen_context(
			context_id, reason or "", active_from, active_to, user=frappe.session.user, expected_version=expected_version
		),
	}
	if action not in dispatch:
		frappe.throw(f"Unknown PE/FY Context action: {action}", frappe.ValidationError)
	return run_idempotent(idempotency_key, "PE Fiscal Year Context", context_id, action, dispatch[action])


# --- Resolution ----------------------------------------------------------------------


@frappe.whitelist()
def resolve_authorized_contexts(remembered_context: str | None = None):
	return resolver.resolve_authorized_contexts(frappe.session.user, remembered_context=remembered_context)


@frappe.whitelist()
def validate_context_for_command(context_id: str):
	return resolver.validate_context_for_command(frappe.session.user, context_id)
